sys/freebsd: add kqueue(2)

tigergao99 opened the pull request.

markjdb commented on sys/freebsd/kqueue.txt:
```diff
@@ -0,0 +1,22 @@
+include <sys/event.h>
+
+kqueue() fd
+kevent(fd fd, changelist ptr[in, array[kevent]], nchanges len[changelist], eventlist ptr[in, array[kevent]], nevents len[eventlist], timeout ptr[in, timespec])
```
Better to define a new resource for kqueues and use that instead: `resource kqueue[fd]`. See the section on resources in the syscall description language docs.

markjdb commented on sys/freebsd/kqueue.txt:
```diff
@@ -0,0 +1,22 @@
+	data	int64
+	udata	intptr
+	ext	array[int64, 4]
+}
```
So in event.h you'll notice that we also have a compatibility structure, kevent_freebsd11. Correspondingly, in syscalls.master there is a COMPAT11 definition for kevent() using struct kevent_freebsd11.

This basically means that during the FreeBSD 11 development cycle we changed the layout of struct kevent. These compat structures exist to allow software compiled using the old struct kevent layout to continue working on FreeBSD without a recompile.

It's important to fuzz these compat system calls as well since the kernel code which handles them by converting from the freebsd11 kevent layout to the current layout can be and frequently is a source of bugs. If you search for "freebsd11" and "freebsd12" in the FreeBSD syscall descriptions you'll see some examples of how to cover compat syscalls.

tigergao99 commented:
Addressed the comments. One more question - I see the kevent32 compat structures defined in event.h. However, I don't see it being used in syscalls in syscalls.master - is there anything we need to do for kevent32 structs?

markjdb commented:
> Addressed the comments. One more question - I see the kevent32 compat structures defined in event.h. However, I don't see it being used in syscalls in syscalls.master - is there anything we need to do for kevent32 structs?

Happily, no. syzkaller is able to fuzz the compat32 layer. It generates an internal representation of the system call interface for each OS architecture that it targets (amd64 and i386 for FreeBSD), and it can fuzz amd64+compat32 by running a fuzzer targeting i386 in a VM running an amd64 kernel. If you look at the FreeBSD syzbot page, you'll two instances: ci-freebsd-i386 is actually targeting a amd64+compat32 kernel.
